internal/provider/resource_application_test.go

```diff
@@ -809,6 +1000,50 @@ func testAccResourceApplicationBasic_Minimal(modelName, charmName string) string
 		`, modelName, charmName)
 }
 
+func testAccResourceApplicationWithExistingModel(appName, username, modelName string) string {
+	return internaltesting.GetStringFromTemplateWithData(
+		"testAccResourceApplicationWithSharedModel",
+		`	
+resource "juju_application" "this" {
+	name = "{{.AppName}}"
+	model = "{{.Username}}/{{.ModelName}}"
+	charm {
+	name = "jameinel-ubuntu-lite"
+	}
+	trust = true
+	expose{}
+}
+		`, internaltesting.TemplateData{
+			"ModelName": modelName,
+			"AppName":   appName,
+			"Username":  username,
+		})
+}
+
+func testAccResourceApplicationWithFullySpecifiedModel(appName, username, modelName string) string {
+	return internaltesting.GetStringFromTemplateWithData(
+		"testAccResourceApplicationWithFullySpecifiedModel",
+		`
+resource "juju_model" "this" {
+	name = "{{.ModelName}}"
+}
+	
+resource "juju_application" "this" {
+	name = "{{.AppName}}"
+	model = "{{.Username}}/${juju_model.this.name}"
```

Comment:
I'm concerned this is a step in the opposite direction from where we want to be with referencing a model inside of other resources, by using the model UUID. The idea is that if we have the model idea, no cache will be necessary. 

Also, looking at this example, you've created a model XYZ for the default user, then are deploying a an application to model XYZ owned by a different user. The linkage inferred by using `juju_model.this.name` appears to be incorrect.

Reply:
I agree that if we don't reference models by name and rather by UUID then this problem goes away. Happy to close this PR in favour of that change but I was aiming for something that maintains backwards compatability.

> Also, looking at this example, you've created a model XYZ for the default user, then are deploying a an application to model XYZ owned by a different user. The linkage inferred by using juju_model.this.name appears to be incorrect.

In this case the "different user" is the same as the "default user", we pass in "admin" who is the user making the connection. If we were to pass in a different user the test should fail with "model not found", I can add that as a test case.
I'm not sure I understand what you mean by "linkage inferred by using juju_model.this.name appears to be incorrect."?